File: android/src/main/kotlin/com/example/pauza_screen_time/usage_stats/model/AppStandbyBucket.kt
package com.example.pauza_screen_time.usage_stats.model

import android.app.usage.UsageStatsManager
import android.os.Build

/**
 * Typed wrapper for Android's app standby bucket constants.
 *
 * Available on API 28+. Each bucket determines how aggressively the system
 * restricts the app's background work and alarms.
 *
 * @see UsageStatsManager.getAppStandbyBucket
 */
enum class AppStandbyBucket(val rawValue: Int) {
    /** App is actively in use. No restrictions. (API 28+) */
    ACTIVE(UsageStatsManager.STANDBY_BUCKET_ACTIVE),

    /** App has been used recently. Light restrictions. (API 28+) */
    WORKING_SET(UsageStatsManager.STANDBY_BUCKET_WORKING_SET),

    /** App is used regularly but not daily. Moderate restrictions. (API 28+) */
    FREQUENT(UsageStatsManager.STANDBY_BUCKET_FREQUENT),

    /** App was last used weeks ago. Heavy restrictions. (API 28+) */
    RARE(UsageStatsManager.STANDBY_BUCKET_RARE),

    /**
     * App was never used or only used briefly. Strictest restrictions. (API 30+)
     *
     * Falls back to [RARE] on API < 30.
     */
    RESTRICTED(if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) 45 else UsageStatsManager.STANDBY_BUCKET_RARE),

    /** Unrecognised value — should not normally occur. */
    UNKNOWN(-1);

    companion object {
        /**
         * Returns the [AppStandbyBucket] matching [rawValue], or [UNKNOWN] if not recognised.
         */
        fun fromRawValue(raw: Int): AppStandbyBucket =
            entries.firstOrNull { it.rawValue == raw } ?: UNKNOWN
    }
}

File: lib/src/features/usage_stats/model/app_standby_bucket.dart
/// Typed representation of Android's app standby bucket values.
///
/// The system assigns each app to a standby bucket that determines how
/// aggressively background work and alarms are restricted.
///
/// **Android 9 (API 28) and above only.**
enum AppStandbyBucket {
  /// App is actively in use. No restrictions.
  active(10),

  /// App has been used recently. Light restrictions.
  workingSet(20),

  /// App is used regularly but not daily. Moderate restrictions.
  frequent(30),

  /// App was last used weeks ago. Heavy restrictions.
  rare(40),

  /// App was never used or used only briefly. Strictest restrictions (API 30+).
  restricted(45),

  /// Unrecognised value returned by the platform — should not normally occur.
  unknown(-1);

  /// The raw integer value returned by the Android platform.
  final int rawValue;

  const AppStandbyBucket(this.rawValue);

  /// Returns the [AppStandbyBucket] for [rawValue], or [unknown] if not recognised.
  static AppStandbyBucket fromRawValue(int value) =>
      values.firstWhere((e) => e.rawValue == value, orElse: () => unknown);
}
